### ISSUE TITLE
Errors in calculation of utility in  MOEA/D algorithm- MOEAD.updateUtility()

### ISSUE THREAD

**Issue by abdulhameedpk** [MOEAFramework/MOEAFramework#10]
There are three mistakes in the calculation of update utility in MOEA/D algorithm provided by  moeaframework. See attached image from source paper [1],

![utility calc source paper](https://cloud.githubusercontent.com/assets/15343255/10762184/c3146154-7ce6-11e5-9959-f6c67e7ac8a4.png)

According to the paper of utility based MOEA/D [1], the utility is updated as follows,
if (relativeDecrease > 0.001)
	π=1
else
	π = ( 0.95 + 0.05(relativeDecrease/0.001) )*π

The snapshot of code of MOEAD.updateUtility() (moeaframework ver 2.4, 2.6) is as follows,  

```
/**
	 * Updates the utility of each individual.
	 */
	protected void updateUtility() {
		for (Individual individual : population) {
			double oldFitness = individual.getFitness();
			double newFitness = fitness(individual.getSolution(), idealPoint);
			double relativeDecrease = oldFitness - newFitness;
		
        	if (relativeDecrease > 0.001) {
				individual.setUtility(1.0);
			} else {
				double utility = Math.min(1.0, 0.95 * (1.0 + delta / 0.001)
						* individual.getUtility());
				individual.setUtility(utility);
			}

			individual.setFitness(newFitness);
		}
	}
```

Now according to the source paper, there are three mistakes in the implementation of MOEA/D.
**Mistak#1**. The relativeDecrease should be,
 
```double relativeDecrease = (oldFitness - newFitness)/ oldFitness;```

**Mistak#2**. Variable delta is used in the else part instead of  ```relativeDecrease```. The variable delta (in moeaframework.MOEAD) is the probability of mating with a solution in the neighborhood rather than the entire population.  

**Mistak#3**.  According to the source paper, if fitness is improved less than 0.001, then utility should be set to  ```(0.95 + 0.05(relativeDecrease/0.001))*utility ```. 
So the correct step is 

```double utility = Math.min(1.0, (0.95 + .05*(relativeDecrease / 0.001) ) * individual.getUtility()); ```

So in my opinion, the correct implementation of MOEAD.updateUtility() is,

```
/**
	 * Updates the utility of each individual.
	 */
	protected void updateUtility() {
		for (Individual individual : population) {
			double oldFitness = individual.getFitness();
			double newFitness = fitness(individual.getSolution(), idealPoint);
			double relativeDecrease = (oldFitness - newFitness)/ oldFitness; 

			if (relativeDecrease > 0.001) {
				individual.setUtility(1.0);
			} else {
				double utility = Math.min(1.0, (0.95 + 0.05*relativeDecrease/0.001)
						* individual.getUtility());
				individual.setUtility(utility);
			}

			individual.setFitness(newFitness);
		}
	}
```

**Note:** The Initial implementation of utility-based MOEA/D[1] provided by the authors (link in [2]) calculate update utility in the same way (in file algorithm.h (CMOEAD::comp_utility()) ) as that of current implementation of moeaframework. The newer implementation (link in [3]) provided by the same author() now corrected the errors. 

References:
[1] Zhang, Q., et al.  "The Performance of a New Version of MOEA/D on  CEC09 Unconstrained MOP Test Instances."  IEEE Congress on Evolutionary  Computation, 2009.
[2] http://dces.essex.ac.uk/staff/zhang/MOEAcompetition/CEC09final/code/ZhangMOEADcode/moead-cppcec2009.zip
[3]http://dces.essex.ac.uk/staff/zhang/MOEAcompetition/CEC09final/code/ZhangMOEADcode/moead_cec2009_2011version.zip

**Comment by dhadka** [comment]
Thanks for spotting this!  We based the Java code off of the original author's first implementation.  Will update the utility calculation for the next release.

**Comment by abdulhameedpk** [comment]
One more correction should be considered. The utility is updated after 50 generations according to the paper[1]. Current implementation of MOEAD implement this in **MOEAD.iterate()** as

```
....
if ((updateUtility >= 0) && (generation % updateUtility == 0)) {
			updateUtility();
		}
...
```
The correct step should be 
```
....
if ((updateUtility >= 0) && (generation % 50 == 0)) {
			updateUtility();
		}
...
```

**Comment by dhadka** [comment]
Thanks for the suggestion.  I'm going to leave the `updateUtility` as-is, since users may want to test the algorithm with different values.  However, I did update the documentation to indicate the recommended value for `updateUtility` is 50.

Both updates have been pushed to master and will be officially released in version 2.7.  Please let us know if you spot any other issues.
